packetio/buffer.go

```diff
@@ -2,88 +2,171 @@
 package packetio
 
 import (
+	"errors"
 	"io"
 	"sync"
 	"time"
 
 	"github.com/pion/transport/deadline"
 )
 
+var errPacketTooBig = errors.New("packet too big")
+
 // Buffer allows writing packets to an intermediate buffer, which can then be read form.
 // This is verify similar to bytes.Buffer but avoids combining multiple writes into a single read.
 type Buffer struct {
-	mutex   sync.Mutex
-	packets [][]byte
+	mutex sync.Mutex
+
+	// this is a circular buffer.  If head <= tail, then the useful
+	// data is in the interval [head, tail[.  If tail < head, then
+	// the useful data is the union of [head, len[ and [0, tail[.
+	// In order to avoid ambiguity when head = tail, we always leave
+	// an unused byte in the buffer.
+	data       []byte
+	head, tail int
 
 	notify chan struct{}
 	subs   bool
 	closed bool
 
-	// The number of buffered packets in bytes.
-	size int
+	count                 int
+	limitCount, limitSize int
 
-	// The limit on Write in packet count and total size.
-	limitCount int
-	limitSize  int
-
-	// Read deadline timer.
 	readDeadline *deadline.Deadline
 }
 
-// NewBuffer creates a new Buffer object.
+const minSize = 2048
+const cutoffSize = 128 * 1024
+const maxSize = 4 * 1024 * 1024
```

Comment:
I think this highly depends on the application and should be configurable from the public API.

Reply:
Since we grow the buffer exponentially, the exact values have a negligible effect on the execution time.  For example, doubling `minSize` only saves a single allocation.

Reply:
In my understanding, grown buffer size is limited by this `maxSize` at https://github.com/pion/transport/pull/72/files#diff-6415ee44ec80ba409c064c7c902b4183R77-R79.
So, an application requiring more than 4MB of the buffer never work any more. Am I missing something?

Reply:
In the previous implementation, it was required for an application to set the maximum size with `SetSizeLimit`.  If it didn't, then it would buffer unbounded amounts of data, which would eventually cause it to crash.

In this version, the size of the buffer is limited by `maxSize`.  This prevents the crash even if the application doesn't call `SetSizeLimit`.

This is a temporary workaround.  The proper fix is to pass the size limit as an argument to `NewBuffer`, so that it's impossible to forget to set the buffer size.  That's for later, though, first I want to get this patch in, which improves both efficiency and robustness.